### PR TITLE
feat: support for submodel sending signals

### PR DESCRIPTION
**Pull request by christopherhuber**
## Context
Search service sends models with nested fields (i.e. serializers submodel fields). CUD operations on sub models do not fire update signal so search data grows stale

## Description
Allows user to add submodels to fire CUD signals when declaring cud event.
Adds post save and post delete signals to each such submodel
On save/delete related target model instances are extracted and update event is fired 

### Subtleties
1 - if found at depth=1 won't keep looking
Currently function won't look for other related instances if target model can be found directly
e.g. 
S -> T1
|-> B -> T2
where S is source, T1, T2 target models and B bridge model
Only T1 will be found

I did this because overwhelming use case is depth one relations and none of our current search serializers require it. But to fix just remove
```
    if target_instances:
        return target_instances
```

2 - could improve performance by more complex input

If instead of taking in list of sub model types, it took in dictionary of sub_models to bridge models; function could avoid unnecessary searching calls - tradeoff is more complexity for the declarer 
e.g.
```
{
    sub_model_1: [],
    sub_model_2: [bridge_model],
    sub_model_3: [bridge_model1, bridge_model2]
}

```

## Additional information (optional)
Open to ideas about best place to store extract_related_instances function

### CODE REVIEW

**Comment by royselfdecode** [comment]
@christopherhuber Please update and resolve the conflicts so we can release this together with the other PR.
Same question: which service will need this update?  Is that only Labs will need. Maybe Reports too because the translations are also related models.

**Comment by royselfdecode** [comment]
@christopherhuber  Also Waseem point out that now we have SNS, it has a different file for declare_cud_event(), please also apply your changes in 
https://github.com/selfdecode/events-library/blob/master/events_library/core/event_bus_sns.py

**Comment by christopherhuber** [comment]
> @christopherhuber Please update and resolve the conflicts so we can release this together with the other PR. Same question: which service will need this update? Is that only Labs will need. Maybe Reports too because the translations are also related models.

Report service definitely needs it - most of recommendation + trait text is included in related report search body. 

Other ones that we have to think about are labs and orders. Both of these have many related models but they only send a slug - usually name or tag. Things that I don't think change that frequently. These are pretty low priority. See list of models  in markers + orders below (you can check fields in question by looking at search_serializer.py files)

- markers (labs)
- lab panels (orders)
- lab tests (orders)
- home test kits (orders)

**Review by royselfdecode — APPROVED** [review]

**Review by waseem-uddin — APPROVED** [review]

File: events_library/core/utils.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+from typing import List, Optional, Set, Type
+
+from django.db.models import Model
+from django.db.models.fields.related import ManyToManyField
+from django.db.models.fields.reverse_related import OneToOneRel
+
+
+def extract_related_instances(
```
Comment (diraven) [inline events_library/core/utils.py:8]: How does this work on the receiving end? Do we need to make a model in the receiving service for each of these submodels?

Comment (christopherhuber) [reply events_library/core/utils.py:8]: No - the receiving end shouldn't require any changes. The event defining service specifies the submodels in the declare event and then submodel cud events get treated as related parent model instances edit events. 

This is designed to be used in conjunction with a custom serializer that sends nested fields. And so to stay up to date requires an updated payload to be sent on cud events of itself or its related sub models.

Comment (diraven) [reply events_library/core/utils.py:8]: Commented in #events-update slack channel.

**Pull request closed without merging**